### PR TITLE
`videoio`: fix `cv::VideoWriter` with FFmpeg encapsulation timestamps

### PR DESCRIPTION
**Pull request by cudawarped**
Fix https://github.com/opencv/opencv/issues/25873 by modifying `cv::VideoWriter` to use provided presentation indices (pts).

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by cudawarped** [comment]
Updated to address warnings

>  /home/ci/opencv/modules/videoio/test/test_ffmpeg.cpp:335:12: warning: suggest explicit braces to avoid ambiguous 'else' [-Wdangling-else]
> 
> /home/ci/opencv/modules/videoio/test/test_ffmpeg.cpp:352:15: warning: suggest explicit braces to avoid ambiguous 'else' [-Wdangling-else]
> 
> /home/ci/opencv/modules/videoio/test/test_ffmpeg.cpp:387:16: warning: suggest explicit braces to avoid ambiguous 'else' [-Wdangling-else]

File: modules/videoio/include/opencv2/videoio.hpp [hunk 1]
```diff
@@ -230,8 +230,10 @@ enum VideoWriterProperties {
...
+  VIDEOWRITER_PROP_KEY_INTERVAL = 10, //!< (**open-only**) Set the key frame interval using raw video encapsulation (\ref VIDEOWRITER_PROP_RAW_VIDEO != 0). Defaults to 1 when not set. FFmpeg back-end only.
+  VIDEOWRITER_PROP_KEY_FLAG = 11, //!< Set to non-zero to signal that the following frames are key frames or zero if not, when encapsulating raw video (\ref VIDEOWRITER_PROP_RAW_VIDEO != 0). FFmpeg back-end only.
+  VIDEOWRITER_PROP_PTS_INDEX = 12, //!< Specifies the frame presentation index for each frame when encapsulating raw video (\ref VIDEOWRITER_PROP_RAW_VIDEO != 0). This flag is necessary when encapsulating externally encoded video where the decoding order differs from the presentation order, such as in GOP patterns with bi-directional B-frames.  FFmpeg back-end only.
+  VIDEOWRITER_PROP_B_FRAME_PRESENTATION_DELAY = 13, //!< Maximum delay (in frames) between a B-frame's decoding and presentation. For example, in a GOP with presentation order IBP and decoding order IPB, this value would be 1, as the B-frame is decoded third but presented second.
```
Comment (asmorkalov) [inline modules/videoio/include/opencv2/videoio.hpp:236]: It'll be great to add some references to `cv::VideoCapture` properties. It's not obvious, if the properties are needed and how to extract them from the encoded stream.

Comment (asmorkalov) [reply modules/videoio/include/opencv2/videoio.hpp:236]: I mean slightly different thing. The VideoWriter user should get values for the new properties somehow. I see 2 sources:
- get something for cv::VideoCapture and promote it to writer (good case). For the case we need a 
- dig in IP camera settings or encoder settings manually (bad case). The case means that we cannot implement generic code that works for all streams.

Comment (cudawarped) [reply modules/videoio/include/opencv2/videoio.hpp:236]: Right I was thinking of streams provided by external encoders (`cudacodec::VideoWriter`) not of the encapsulation of RTSP streams.

For the first parameter (`VIDEOWRITER_PROP_PTS_INDEX`), if the frame rate is provided in the rtsp stream it is equivalent to dividing the current position of the video file (`CAP_PROP_POS_MSEC`) by the duration of the frame (i.e. 1000.0/`VideoCapture::get(CAP_PROP_FPS)` if the frame rate is fixed.  If the frame rate isn't fixed then we would need to create an extra property inside `cv::VideoCapture` to provide this information?

I am not sure if the second parameter (`VIDEOWRITER_PROP_B_FRAME_PRESENTATION_DELAY`) is available inside FFmpeg but if it is (for all versions) we would need to create this extra property inside `cv::VideoCapture` as well.

Comment (cudawarped) [reply modules/videoio/include/opencv2/videoio.hpp]: @asmorkalov I've updated VideoCapture to expose the required properties.

**Comment by asmorkalov** [comment]
Linux warning:
```
Run cd $HOME/build && python3 $HOME/scripts/warnings-handling.py
/home/ci/opencv/modules/videoio/src/cap_ffmpeg_impl.hpp:1873:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
```

**Comment by cudawarped** [comment]
> Linux warning:
> 
> ```
> Run cd $HOME/build && python3 $HOME/scripts/warnings-handling.py
> /home/ci/opencv/modules/videoio/src/cap_ffmpeg_impl.hpp:1873:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
> ```

I've pushed an update which should address this.

**Comment by asmorkalov** [comment]
@opencv-alalek @mshabunin could you take a look too?

**Review by mshabunin — APPROVED** [review]

**Comment by asmorkalov** [comment]
Reporduced failed test with Ubuntu 16.04 locally. Will take a look, if we can handle this:
```
ubuntu@jetson1:~/Projects/opencv-build$ ./bin/opencv_test_videoio --gtest_filter=videoio_encapsulate.write/16
CTEST_FULL_OUTPUT
OpenCV version: 4.10.0-dev
OpenCV VCS version: 4.10.0-98-geafbf4b-dirty
Build type: Release
Compiler: /usr/bin/c++  (ver 5.4.0)
Parallel framework: pthreads (nthreads=1)
CPU features: NEON
OpenCL is disabled
TEST: Skip tests with tags: 'mem_2gb', 'verylong'
Note: Google Test filter = videoio_encapsulate.write/16
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from videoio_encapsulate
[ RUN      ] videoio_encapsulate.write/16, where GetParam() = ("video/sample_322x242_15frames.yuv420p.libx264.mp4", "mp4", 15, 15, true, true)
OpenCV: FFMPEG: tag 0x34363268/'h264' is not supported with codec id 28 and format 'mp4 / MP4 (MPEG-4 Part 14)'
OpenCV: FFMPEG: fallback to use tag 0x00000021/'!???'
Checking encapsulated video container: /tmp/__opencv_temp.OzpE2c.test_encapsulated_stream.mp4
[NULL @ 0x17c550] Packet header is not contained in global extradata, corrupted stream or invalid MP4/AVCC bitstream
/home/ubuntu/Projects/opencv/modules/videoio/test/test_ffmpeg.cpp:380: Failure
Value of: capActualRaw.grab()
  Actual: false
Expected: true
[  FAILED  ] videoio_encapsulate.write/16, where GetParam() = ("video/sample_322x242_15frames.yuv420p.libx264.mp4", "mp4", 15, 15, true, true) (77 ms)
[----------] 1 test from videoio_encapsulate (78 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (80 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] videoio_encapsulate.write/16, where GetParam() = ("video/sample_322x242_15frames.yuv420p.libx264.mp4", "mp4", 15, 15, true, true)
```

**Comment by asmorkalov** [comment]
Looks similar to what we discussed in https://github.com/opencv/opencv/pull/24363.

**Comment by cudawarped** [comment]
> Looks similar to what we discussed in #24363.

Its the same issue but unfortunately can't be fixed by using a *.avi* container because the pts infomation is not available.  Changing to *.ts* works for me with FFMpeg 2.8.22 (Ubuntu 16.04) and 4.2.9 (Ubuntu 20.04)

**Pull request merged**